packet/src/bwe/rate_controller.rs

```diff
@@ -0,0 +1,376 @@
+    min_bitrate: f64,
+    max_bitrate: f64,
+
+    /// The last observed bitrate calculated based on TWCC data.
+    last_observed_bitrate: Option<f64>,
+    /// The averaged observed bitrate when we have been in the decrease state.
+    averaged_observed_bitrate: MovingAverage,
+    /// The last time we updated the estimated bitrate.
+    last_estimate_update: Instant,
+    // Last RTT estimate in micro-seconds
+    last_rtt_us: Option<f64>,
+}
+
+impl RateController {
+    pub(super) fn new(
+        estimated_bitrate: u64,
+        min_bitrate: u64,
+        max_bitrate: u64,
+        now: Instant,
+    ) -> Self {
+        Self {
+            state: State::Increase,
+
+            estimated_bitrate: estimated_bitrate as f64,
+            min_bitrate: min_bitrate as f64,
+            max_bitrate: max_bitrate as f64,
+
+            last_observed_bitrate: None,
+            averaged_observed_bitrate: MovingAverage::new(OBSERVED_BIT_RATE_SMOOTHING_FACTOR),
+            last_estimate_update: now,
+            last_rtt_us: None,
+        }
+    }
+
+    /// Update the observed bitrate received by the receiver.
+    ///
+    /// This is typically based on summarising TWCC feedback and correlating with known send sizes
+    /// over a window.
+    pub(super) fn update_observed_bitrate(&mut self, observed_bitrate: u64) {
+        self.last_observed_bitrate = Some(observed_bitrate as f64);
+    }
+
+    /// Update the estimated round trip time(from TWCC or RR).
+    pub(super) fn update_rtt(&mut self, rtt_us: f64) {
+        self.last_rtt_us = Some(rtt_us);
+    }
+
+    /// Update with input from the delay controller.
+    pub(super) fn update_delay(&mut self, signal: Signal, now: Instant) {
+        self.state = self.state.transition(signal);
+
+        match self.state {
+            State::Increase => {
+                self.increase(now);
+            }
+            State::Decrease => {
+                self.decrease(now);
+            }
+            State::Hold => {
+                // Do nothing
+            }
+        }
+    }
+
+    pub(super) fn estimated_bitrate(&self) -> u64 {
+        self.estimated_bitrate.floor() as u64
+    }
+
+    fn increase(&mut self, now: Instant) {
+        let since_last_update = ((now - self.last_estimate_update).as_millis() as f64) / 1000.0;
+
+        let new_estimate = if self.is_near_convergence() {
+            // Additive increase
+            let response_time_ms = self.last_rtt_us.map(|rtt| rtt / 1000.0).unwrap_or(0.0) + 100.0;
+            let alpha = 0.5 * (since_last_update / response_time_ms).min(1.0);
+            let expected_packet_size = self.estimated_packet_size();
+            self.estimated_bitrate + (alpha * expected_packet_size).max(1000.0)
+        } else {
+            // Multiplicative increase
+            let eta = 1.08_f64.powf(since_last_update.min(1.0));
+            eta * self.estimated_bitrate
+        };
+        let max = self.last_observed_bitrate.map(|r| r * 1.5_f64);
```

Comment:
I think a lot of these constants should be given constant names (like BETA and OBSERVED_BIT_RATE_SMOOTHING_FACTOR).